zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java

```diff
@@ -306,6 +329,23 @@ public Interpreter getCurrentRepl() {
     return completion;
   }
 
+  public int calculateCursorPosition(String buffer, int cursor) {
+    int countWhitespacesAtStart = buffer.indexOf(buffer.trim());
+    if (countWhitespacesAtStart > 0) {
+      cursor -= countWhitespacesAtStart;
+    }
+
+    buffer = buffer.trim();
+
+    String replName = getRequiredReplName(buffer);
+    if (replName != null && cursor > replName.length()) {
```

Comment:
nit: there's a fair bit of overlap here. here and `completion()`, `buffer.trim()` and `getRequireReplName(buffer)` is called x2